File: server/src/signaling.js
// WebSocket signaling: room registry + SDP/ICE relay (docs/PROTOCOL.md §2, TR2).
// The server never parses SDP — it routes on (room from the socket's session, peerId).
import { randomInt, randomUUID } from 'node:crypto';

// 6-char room codes from A–Z/2–9 excluding ambiguous 0/O/1/I (PROTOCOL §2.1).
const ROOM_CODE_ALPHABET = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
const ROOM_CODE_LENGTH = 6;
const MAX_PARENTS = 4;
const ROOM_GRACE_MS = 10 * 60 * 1000; // rooms GC'd 10 min after both sides gone

const OPEN = 1; // WebSocket.OPEN

function isOpen(ws) {
  return ws && ws.readyState === OPEN;
}

function sendJson(ws, obj) {
  if (isOpen(ws)) ws.send(JSON.stringify(obj));
}

function sendError(ws, code, message) {
  sendJson(ws, { type: 'error', code, message });
}

export function createSignaling({ gcIntervalMs = 60_000 } = {}) {
  /** roomId -> { roomId, reclaimToken, camera, parents: Map<peerId, ws>,
   *              parentAuth: Map<peerId, auth>, emptySince } */
  const rooms = new Map();

  function generateRoomId() {
    let id;
    do {
      id = '';
      for (let i = 0; i < ROOM_CODE_LENGTH; i++) {
        id += ROOM_CODE_ALPHABET[randomInt(ROOM_CODE_ALPHABET.length)];
      }
    } while (rooms.has(id));
    return id;
  }

  function maybeMarkEmpty(room) {
    if (!isOpen(room.camera) && room.parents.size === 0) {
      room.emptySince = Date.now();
    }
  }

  /** Remove rooms where both sides have been gone for >= 10 minutes. */
  function gcNow(now = Date.now()) {
    for (const [roomId, room] of rooms) {
      if (room.emptySince !== null && now - room.emptySince >= ROOM_GRACE_MS) {
        rooms.delete(roomId);
      }
    }
  }

  const gcTimer = setInterval(gcNow, gcIntervalMs);
  gcTimer.unref?.();

  /** Detach ws from its room (used for both 'leave' and socket close). */
  function leaveRoom(ws, ctx) {
    const room = ctx.room;
    if (!room) return;
    if (ctx.role === 'camera') {
      if (room.camera === ws) {
        room.camera = null;
        // Parents show RECONNECTING and wait; room persists for reclaim (§2.1).
        for (const parentWs of room.parents.values()) {
          sendJson(parentWs, { type: 'camera-left' });
        }
        maybeMarkEmpty(room);
      }
    } else if (ctx.role === 'parent') {
      if (room.parents.get(ctx.peerId) === ws) {
        room.parents.delete(ctx.peerId);
        room.parentAuth.delete(ctx.peerId);
        sendJson(room.camera, { type: 'peer-left', peerId: ctx.peerId });
        maybeMarkEmpty(room);
      }
    }
    ctx.room = null;
    ctx.role = null;
    ctx.peerId = null;
  }

  function handleCreateRoom(ws, ctx, msg) {
    if (ctx.room) leaveRoom(ws, ctx); // a socket holds at most one room session

    if (msg.roomId !== undefined || msg.reclaimToken !== undefined) {
      // Reclaim path: camera reconnected and wants its old room back.
      const room = rooms.get(msg.roomId);
      if (!room || typeof msg.reclaimToken !== 'string' || room.reclaimToken !== msg.reclaimToken) {
        sendError(ws, 'BAD_RECLAIM', 'Unknown room or wrong reclaim token');
        return;
      }
      if (room.camera && room.camera !== ws) room.camera.terminate?.();
      room.camera = ws;
      room.emptySince = null;
      ctx.role = 'camera';
      ctx.room = room;
      sendJson(ws, { type: 'room-created', roomId: room.roomId, reclaimToken: room.reclaimToken });
      // Re-announce live parents so the camera re-creates a PC + offer per peer.
      // Carry the auth captured at join time (§8.2) so the camera re-runs its
      // trusted-device challenge for each re-announced parent.
      for (const [peerId, parentWs] of room.parents) {
        if (!isOpen(parentWs)) continue;
        const peerJoined = { type: 'peer-joined', peerId };
        const savedAuth = room.parentAuth.get(peerId);
        if (savedAuth != null) peerJoined.auth = savedAuth;
        sendJson(ws, peerJoined);
      }
      return;
    }

    const room = {
      roomId: generateRoomId(),
      reclaimToken: randomUUID(),
      camera: ws,
      parents: new Map(),
      parentAuth: new Map(),
      emptySince: null,
    };
    rooms.set(room.roomId, room);
    ctx.role = 'camera';
    ctx.room = room;
    sendJson(ws, { type: 'room-created', roomId: room.roomId, reclaimToken: room.reclaimToken });
  }

  function handleJoinRoom(ws, ctx, msg) {
    if (typeof msg.roomId !== 'string') {
      sendError(ws, 'BAD_MESSAGE', 'join-room requires roomId');
      return;
    }
    if (ctx.room) leaveRoom(ws, ctx);

    const room = rooms.get(msg.roomId);
    if (!room) {
      sendError(ws, 'ROOM_NOT_FOUND', `No room ${msg.roomId}`);
      return;
    }

    // Optional peerId reuse on rejoin: honoured when free (not held by a live socket).
    let peerId = typeof msg.peerId === 'string' && msg.peerId !== '' ? msg.peerId : null;
    if (peerId !== null) {
      const existing = room.parents.get(peerId);
      if (existing && existing !== ws && isOpen(existing)) {
        peerId = null; // taken by a live parent — treat as a fresh join
      } else if (existing && existing !== ws) {
        existing.terminate?.(); // stale socket; reclaim the slot
      }
    }
    if ((peerId === null || !room.parents.has(peerId)) && room.parents.size >= MAX_PARENTS) {
      sendError(ws, 'ROOM_FULL', `Room ${room.roomId} already has ${MAX_PARENTS} parents`);
      return;
    }
    if (peerId === null) peerId = randomUUID();

    room.parents.set(peerId, ws);
    // Optional trusted-device auth (§8.2): passed through verbatim, contents not
    // validated by the server. Captured per peerId so it survives a camera reclaim.
    const auth = msg.auth;
    if (auth != null) room.parentAuth.set(peerId, auth);
    else room.parentAuth.delete(peerId);
    room.emptySince = null;
    ctx.role = 'parent';
    ctx.room = room;
    ctx.peerId = peerId;
    sendJson(ws, { type: 'room-joined', roomId: room.roomId, peerId });
    // Camera responds with an auth-challenge when auth is present, else an offer (§2.1).
    const peerJoined = { type: 'peer-joined', peerId };
    if (auth != null) peerJoined.auth = auth;
    sendJson(room.camera, peerJoined);
  }

  /** offer/answer/ice/ice-restart/auth-challenge/auth-response: relayed verbatim,
   *  routed on peerId (§2.2). auth-challenge flows camera→parent like offer;
   *  auth-response flows parent→camera like answer (peerId stamped server-side,
   *  the deviceId/pk/sig payload left untouched). */
  function relayNegotiation(ws, ctx, msg, raw) {
    const room = ctx.room;
    if (!room) {
      sendError(ws, 'NOT_IN_ROOM', `${msg.type} requires an active room session`);
      return;
    }
    if (ctx.role === 'camera') {
      const target = room.parents.get(msg.peerId);
      if (isOpen(target)) target.send(raw);
    } else {
      // Stamp the sender's real peerId so the camera routes to the right PC.
      sendJson(room.camera, { ...msg, peerId: ctx.peerId });
    }
  }

  /** hb + noise: camera → all parents fanout (§2.3). */
  function fanout(ws, ctx, msg, raw) {
    const room = ctx.room;
    if (!room) {
      sendError(ws, 'NOT_IN_ROOM', `${msg.type} requires an active room session`);
      return;
    }
    if (ctx.role !== 'camera') return; // only the camera emits hb/noise
    for (const parentWs of room.parents.values()) {
      if (isOpen(parentWs)) parentWs.send(raw);
    }
  }

  function handleMessage(ws, ctx, msg, raw) {
    switch (msg.type) {
      case 'create-room':
        handleCreateRoom(ws, ctx, msg);
        break;
      case 'join-room':
        handleJoinRoom(ws, ctx, msg);
        break;
      case 'offer':
      case 'answer':
      case 'ice':
      case 'ice-restart':
      case 'auth-challenge':
      case 'auth-response':
        relayNegotiation(ws, ctx, msg, raw);
        break;
      case 'hb':
      case 'noise':
        fanout(ws, ctx, msg, raw);
        break;
      case 'leave':
        leaveRoom(ws, ctx);
        break;
      default:
        // Unknown message types MUST be ignored (forward compatibility, NTR6).
        break;
    }
  }

  function handleConnection(ws) {
    const ctx = { role: null, room: null, peerId: null };
    ws.on('message', (data) => {
      const raw = data.toString(); // always relay as text frames
      let msg;
      try {
        msg = JSON.parse(raw);
      } catch {
        sendError(ws, 'BAD_MESSAGE', 'Frame is not valid JSON');
        return;
      }
      if (msg === null || typeof msg !== 'object' || typeof msg.type !== 'string') {
        sendError(ws, 'BAD_MESSAGE', 'Frame must be a JSON object with a string "type"');
        return;
      }
      try {
        handleMessage(ws, ctx, msg, raw);
      } catch {
        sendError(ws, 'BAD_MESSAGE', 'Failed to handle message');
      }
    });
    ws.on('close', () => leaveRoom(ws, ctx));
    ws.on('error', () => {
      /* close event follows; never crash the process (NTR3) */
    });
  }

  return {
    handleConnection,
    gcNow,
    rooms,
    close() {
      clearInterval(gcTimer);
    },
  };
}

File: server-java/src/main/java/com/babymonitor/Signaling.java
package com.babymonitor;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.javalin.websocket.WsContext;

import java.security.SecureRandom;
import java.time.Duration;
import java.time.Instant;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * WebSocket signaling: room registry + SDP/ICE relay (docs/PROTOCOL.md §2, TR2).
 * The server never parses SDP — it routes on (room from the socket's session, peerId).
 *
 * Faithful port of server/src/signaling.js. All handlers run under a single lock,
 * reproducing the Node event-loop's serialized semantics.
 */
public final class Signaling {

    // 6-char room codes from A–Z/2–9 excluding ambiguous 0/O/1/I (PROTOCOL §2.1).
    private static final String ROOM_CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    private static final int ROOM_CODE_LENGTH = 6;
    private static final int MAX_PARENTS = 4;
    private static final long ROOM_GRACE_MS = 10 * 60 * 1000; // rooms GC'd 10 min after both sides gone

    private static final ObjectMapper MAPPER = new ObjectMapper()
            .enable(DeserializationFeature.FAIL_ON_TRAILING_TOKENS);
    private static final SecureRandom RANDOM = new SecureRandom();

    /** Per-connection session state (the Node code's per-socket ctx object). */
    private static final class Conn {
        final WsContext ws;
        String role;    // "camera" | "parent" | null
        Room room;      // null when not in a room
        String peerId;  // parents only
        JsonNode auth;  // parents only: trusted-device auth captured at join (§8.2), or null

        Conn(WsContext ws) {
            this.ws = ws;
        }
    }

    /** roomId -> { roomId, reclaimToken, camera, parents, emptySince } */
    static final class Room {
        final String roomId;
        final String reclaimToken;
        Conn camera;
        final LinkedHashMap<String, Conn> parents = new LinkedHashMap<>();
        Long emptySince; // epoch ms; null while either side is present

        Room(String roomId, String reclaimToken) {
            this.roomId = roomId;
            this.reclaimToken = reclaimToken;
        }
    }

    private final Map<String, Room> rooms = new LinkedHashMap<>();
    private final Map<String, Conn> conns = new HashMap<>(); // keyed by WS session id
    private final ScheduledExecutorService gcTimer;

    public Signaling() {
        this(60_000);
    }

    public Signaling(long gcIntervalMs) {
        gcTimer = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "signaling-gc");
            t.setDaemon(true); // like Node's timer.unref()
            return t;
        });
        gcTimer.scheduleAtFixedRate(() -> gcNow(Instant.now()), gcIntervalMs, gcIntervalMs,
                TimeUnit.MILLISECONDS);
    }

    // ------------------------------------------------------------------ WS event entry points

    public synchronized void onConnect(WsContext ctx) {
        // Signaling connections are long-lived and may be receive-only on the parent
        // side; disable Jetty's idle timeout like the Node server (which has none).
        try {
            ctx.session.setIdleTimeout(Duration.ZERO);
        } catch (Exception ignored) {
            // best effort; hb traffic keeps active rooms alive anyway
        }
        conns.put(ctx.sessionId(), new Conn(ctx));
    }

    public synchronized void onMessage(WsContext ctx, String raw) {
        Conn conn = conns.computeIfAbsent(ctx.sessionId(), id -> new Conn(ctx));
        JsonNode msg;
        try {
            msg = MAPPER.readTree(raw);
        } catch (Exception e) {
            sendError(conn, "BAD_MESSAGE", "Frame is not valid JSON");
            return;
        }
        if (msg == null || !msg.isObject() || !msg.path("type").isTextual()) {
            sendError(conn, "BAD_MESSAGE", "Frame must be a JSON object with a string \"type\"");
            return;
        }
        try {
            handleMessage(conn, (ObjectNode) msg, raw);
        } catch (Exception e) {
            sendError(conn, "BAD_MESSAGE", "Failed to handle message");
        }
    }

    public synchronized void onClose(WsContext ctx) {
        Conn conn = conns.remove(ctx.sessionId());
        if (conn != null) leaveRoom(conn);
    }

    /** Remove rooms where both sides have been gone for >= 10 minutes. */
    public synchronized void gcNow(Instant now) {
        long nowMs = now.toEpochMilli();
        Iterator<Map.Entry<String, Room>> it = rooms.entrySet().iterator();
        while (it.hasNext()) {
            Room room = it.next().getValue();
            if (room.emptySince != null && nowMs - room.emptySince >= ROOM_GRACE_MS) {
                it.remove();
            }
        }
    }

    public synchronized void close() {
        gcTimer.shutdownNow();
    }

    // ------------------------------------------------------------------ test hooks

    public synchronized boolean hasRoom(String roomId) {
        return rooms.containsKey(roomId);
    }

    /** @return epoch ms since the room emptied, or null if absent/occupied */
    public synchronized Long roomEmptySince(String roomId) {
        Room room = rooms.get(roomId);
        return room == null ? null : room.emptySince;
    }

    // ------------------------------------------------------------------ message dispatch

    private void handleMessage(Conn conn, ObjectNode msg, String raw) {
        switch (msg.get("type").asText()) {
            case "create-room" -> handleCreateRoom(conn, msg);
            case "join-room" -> handleJoinRoom(conn, msg);
            case "offer", "answer", "ice", "ice-restart", "auth-challenge", "auth-response" ->
                    relayNegotiation(conn, msg, raw);
            case "hb", "noise" -> fanout(conn, msg, raw);
            case "leave" -> leaveRoom(conn);
            default -> {
                // Unknown message types MUST be ignored (forward compatibility, NTR6).
            }
        }
    }

    private void handleCreateRoom(Conn conn, ObjectNode msg) {
        if (conn.room != null) leaveRoom(conn); // a socket holds at most one room session

        if (msg.has("roomId") || msg.has("reclaimToken")) {
            // Reclaim path: camera reconnected and wants its old room back.
            String roomId = msg.path("roomId").isTextual() ? msg.get("roomId").asText() : null;
            Room room = roomId == null ? null : rooms.get(roomId);
            JsonNode token = msg.get("reclaimToken");
            if (room == null || token == null || !token.isTextual()
                    || !room.reclaimToken.equals(token.asText())) {
                sendError(conn, "BAD_RECLAIM", "Unknown room or wrong reclaim token");
                return;
            }
            Conn old = room.camera;
            room.camera = conn;
            room.emptySince = null;
            conn.role = "camera";
            conn.room = room;
            if (old != null && old != conn) terminate(old);
            ObjectNode created = MAPPER.createObjectNode();
            created.put("type", "room-created");
            created.put("roomId", room.roomId);
            created.put("reclaimToken", room.reclaimToken);
            sendJson(conn, created);
            // Re-announce live parents so the camera re-creates a PC + offer per peer.
            // Carry the auth captured at join time (§8.2) so the camera re-runs its
            // trusted-device challenge for each re-announced parent.
            for (Map.Entry<String, Conn> entry : room.parents.entrySet()) {
                Conn parent = entry.getValue();
                if (isOpen(parent)) {
                    ObjectNode joined = MAPPER.createObjectNode();
                    joined.put("type", "peer-joined");
                    joined.put("peerId", entry.getKey());
                    if (parent.auth != null) joined.set("auth", parent.auth);
                    sendJson(conn, joined);
                }
            }
            return;
        }

        Room room = new Room(generateRoomId(), UUID.randomUUID().toString());
        room.camera = conn;
        rooms.put(room.roomId, room);
        conn.role = "camera";
        conn.room = room;
        ObjectNode created = MAPPER.createObjectNode();
        created.put("type", "room-created");
        created.put("roomId", room.roomId);
        created.put("reclaimToken", room.reclaimToken);
        sendJson(conn, created);
    }

    private void handleJoinRoom(Conn conn, ObjectNode msg) {
        if (!msg.path("roomId").isTextual()) {
            sendError(conn, "BAD_MESSAGE", "join-room requires roomId");
            return;
        }
        if (conn.room != null) leaveRoom(conn);

        Room room = rooms.get(msg.get("roomId").asText());
        if (room == null) {
            sendError(conn, "ROOM_NOT_FOUND", "No room " + msg.get("roomId").asText());
            return;
        }

        // Optional peerId reuse on rejoin: honoured when free (not held by a live socket).
        String peerId = msg.path("peerId").isTextual() && !msg.get("peerId").asText().isEmpty()
                ? msg.get("peerId").asText()
                : null;
        Conn stale = null;
        if (peerId != null) {
            Conn existing = room.parents.get(peerId);
            if (existing != null && existing != conn && isOpen(existing)) {
                peerId = null; // taken by a live parent — treat as a fresh join
            } else if (existing != null && existing != conn) {
                stale = existing; // stale socket; reclaim the slot
            }
        }
        if ((peerId == null || !room.parents.containsKey(peerId))
                && room.parents.size() >= MAX_PARENTS) {
            sendError(conn, "ROOM_FULL",
                    "Room " + room.roomId + " already has " + MAX_PARENTS + " parents");
            return;
        }
        if (peerId == null) peerId = UUID.randomUUID().toString();

        room.parents.put(peerId, conn);
        if (stale != null) terminate(stale);
        // Optional trusted-device auth (§8.2): passed through verbatim, contents not
        // validated by the server. Captured on the parent's Conn so it survives a reclaim.
        JsonNode authField = msg.get("auth");
        conn.auth = (authField != null && !authField.isNull()) ? authField.deepCopy() : null;
        room.emptySince = null;
        conn.role = "parent";
        conn.room = room;
        conn.peerId = peerId;
        ObjectNode joined = MAPPER.createObjectNode();
        joined.put("type", "room-joined");
        joined.put("roomId", room.roomId);
        joined.put("peerId", peerId);
        sendJson(conn, joined);
        // Camera responds with an auth-challenge when auth is present, else an offer (§2.1).
        ObjectNode peerJoined = MAPPER.createObjectNode();
        peerJoined.put("type", "peer-joined");
        peerJoined.put("peerId", peerId);
        if (conn.auth != null) peerJoined.set("auth", conn.auth);
        sendJson(room.camera, peerJoined);
    }

    /**
     * offer/answer/ice/ice-restart/auth-challenge/auth-response: relayed verbatim,
     * routed on peerId (§2.2). auth-challenge flows camera→parent like offer;
     * auth-response flows parent→camera like answer (peerId stamped server-side,
     * the deviceId/pk/sig payload left untouched).
     */
    private void relayNegotiation(Conn conn, ObjectNode msg, String raw) {
        Room room = conn.room;
        if (room == null) {
            sendError(conn, "NOT_IN_ROOM",
                    msg.get("type").asText() + " requires an active room session");
            return;
        }
        if ("camera".equals(conn.role)) {
            Conn target = msg.path("peerId").isTextual()
                    ? room.parents.get(msg.get("peerId").asText())
                    : null;
            if (isOpen(target)) sendRaw(target, raw);
        } else {
            // Stamp the sender's real peerId so the camera routes to the right PC
            // (a parent cannot spoof another peer).
            ObjectNode stamped = msg.deepCopy();
            stamped.put("peerId", conn.peerId);
            sendJson(room.camera, stamped);
        }
    }

    /** hb + noise: camera → all parents fanout (§2.3). */
    private void fanout(Conn conn, ObjectNode msg, String raw) {
        Room room = conn.room;
        if (room == null) {
            sendError(conn, "NOT_IN_ROOM",
                    msg.get("type").asText() + " requires an active room session");
            return;
        }
        if (!"camera".equals(conn.role)) return; // only the camera emits hb/noise
        for (Conn parent : room.parents.values()) {
            if (isOpen(parent)) sendRaw(parent, raw);
        }
    }

    /** Detach a connection from its room (used for both 'leave' and socket close). */
    private void leaveRoom(Conn conn) {
        Room room = conn.room;
        if (room == null) return;
        if ("camera".equals(conn.role)) {
            if (room.camera == conn) {
                room.camera = null;
                // Parents show RECONNECTING and wait; room persists for reclaim (§2.1).
                ObjectNode cameraLeft = MAPPER.createObjectNode();
                cameraLeft.put("type", "camera-left");
                for (Conn parent : room.parents.values()) {
                    sendJson(parent, cameraLeft);
                }
                maybeMarkEmpty(room);
            }
        } else if ("parent".equals(conn.role)) {
            if (room.parents.get(conn.peerId) == conn) {
                room.parents.remove(conn.peerId);
                ObjectNode peerLeft = MAPPER.createObjectNode();
                peerLeft.put("type", "peer-left");
                peerLeft.put("peerId", conn.peerId);
                sendJson(room.camera, peerLeft);
                maybeMarkEmpty(room);
            }
        }
        conn.room = null;
        conn.role = null;
        conn.peerId = null;
    }

    private void maybeMarkEmpty(Room room) {
        if (!isOpen(room.camera) && room.parents.isEmpty()) {
            room.emptySince = System.currentTimeMillis();
        }
    }

    private String generateRoomId() {
        String id;
        do {
            StringBuilder sb = new StringBuilder(ROOM_CODE_LENGTH);
            for (int i = 0; i < ROOM_CODE_LENGTH; i++) {
                sb.append(ROOM_CODE_ALPHABET.charAt(RANDOM.nextInt(ROOM_CODE_ALPHABET.length())));
            }
            id = sb.toString();
        } while (rooms.containsKey(id));
        return id;
    }

    // ------------------------------------------------------------------ socket helpers

    private static boolean isOpen(Conn conn) {
        try {
            return conn != null && conn.ws.session.isOpen();
        } catch (Exception e) {
            return false;
        }
    }

    private static void sendRaw(Conn conn, String text) {
        if (!isOpen(conn)) return;
        try {
            conn.ws.send(text);
        } catch (Exception ignored) {
            // peer vanished mid-send; close event follows — never crash (NTR3)
        }
    }

    private static void sendJson(Conn conn, ObjectNode obj) {
        sendRaw(conn, obj.toString());
    }

    private static void sendError(Conn conn, String code, String message) {
        ObjectNode error = MAPPER.createObjectNode();
        error.put("type", "error");
        error.put("code", code);
        error.put("message", message);
        sendJson(conn, error);
    }

    /** Abruptly close a superseded socket (the Node code's ws.terminate()). */
    private static void terminate(Conn conn) {
        try {
            conn.ws.closeSession();
        } catch (Exception ignored) {
            // already gone
        }
    }
}
